### PR TITLE
HITL AppState refactor

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

Refactor to better support multiple HITL-tool use cases: rearrange, SocialNav, fetch.

### Limitations
Free camera and tutorial functionality are temporarily unsupported. We'll create new AppStates for these soon.

### AppState classes
SandboxDriver loads a habitat env, including scene/episode/task/agent/policy. A derived AppState class drives use-case-specific behavior for the HITL tool during each env step, including processing user input and manipulating the sim state. In the future, SandboxDriver may manage the transition between AppStates, e.g. transition from a non-interactive tutorial state to a user-controlled-agent state. So far, I have `AppStateRearrange` (the existing HITL functionality) and a work-in-progress `AppStateFetch`. They differ slightly, for example, in the fetch app state, the user isn't expected to move objects to certain goal positions, and the episode doesn't end.

### SandboxService
This is a set of singletons and helpers provided by SandboxDriver to AppStates and other helper classes. The pattern here is sometimes called a "DI Container" (DI = dependency injection).

### GuiNavigationHelper, CameraHelper, hablab_utils.py
These are examples of helpers that allow us to share code/behavior between AppStates. Notably, we use composition here: each AppState object has a GuiNavigationHelper object, etc. Contrast this to dumping all this shared code into the AppState base class; individual helper classes are a cleaner approach because (1) they each encapsulate a specific functionality and (2) derived AppState classes don't inherit base functionality they don't need (e.g. `AppStateFreeCamera` won't construct a `GuiNavigationHelper`).

## How Has This Been Tested

Local testing on Fedora, e.g.
```
HABITAT_SIM_LOG=warning MAGNUM_LOG=warning \
python examples/siro_sandbox/sandbox_app.py \
--app-state fetch \
--disable-inverse-kinematics \
--never-end \
--gui-controlled-agent-index 1 \
--cfg experiments_hab3/pop_play_kinematic_oracle_humanoid_spot_fp.yaml \
--cfg-opts \
habitat_baselines.evaluate=True \
habitat_baselines.num_environments=1 \
habitat_baselines.eval.should_load_ckpt=False \
~habitat.task.measurements.agent_blame_measure
```
## Types of changes

PR into SIRo branch

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes if required.

### CODE REVIEW

File: examples/siro_sandbox/app_states/app_state_fetch.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                GuiInput.KeyNS.SPACE
+            ):
+                # temp: drop object right where it is
+                drop_pos = self._get_target_object_position(
+                    self._held_target_obj_idx
+                )
+                drop_pos.y = 0.0
+                self._held_target_obj_idx = None
+
+            # todo: implement throwing, including viz
```
Comment (eundersander) [inline examples/siro_sandbox/app_states/app_state_fetch.py:79]: @xavierpuigf you can start plugging in your real fetch task in this file.

File: examples/siro_sandbox/sandbox_app.py [hunk 2]
```diff
@@ -140,59 +128,64 @@ def __init__(self, args, config, gui_input):
             self.gym_habitat_env, config, args, gui_input, self._step_recorder
         )
 
-        self.gui_agent_ctrl = self.ctrl_helper.get_gui_agent_controller()
-
-        self.cam_transform = None
-        self.cam_zoom_dist = 1.0
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:146]: These are examples of member variables related to our mouse-controlled camera functionality. All of this is now encapsulated in `camera_helper.py`.

File: examples/siro_sandbox/sandbox_app.py [hunk 3]
```diff
@@ -140,59 +128,64 @@ def __init__(self, args, config, gui_input):
...
-                -min(max(np.radians(args.min_look_down_angle), -np.pi / 2), 0)
-                - 1e-5
+
+        self._viz_anim_fraction = 0.0
+        self._pending_cursor_style = None
+
+        def local_end_episode(do_reset=False):
+            self._end_episode(do_reset)
+
+        self._sandbox_service = SandboxService(
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:152]: This is roughly the "DI container" software pattern.

File: examples/siro_sandbox/sandbox_app.py [hunk 4]
```diff
@@ -364,424 +308,10 @@ def _end_episode(self, do_reset=False):
...
-        self._text_drawer = text_drawer
-
-    def is_free_camera_mode(self):
-        return self.ctrl_helper.get_gui_controlled_agent_index() is None
-
     # trying to get around mypy complaints about missing sim attributes
     def get_sim(self) -> Any:
         return self.habitat_env.task._sim
 
-    def _draw_nav_hint_from_agent(self, end_pos, end_radius, color):
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:395]: The big chunks of code removed from `sandbox_app.py` have mostly been moved to `app_state_rearrange.py`, `gui_navigation_helper.py`, or other helpers.

File: examples/siro_sandbox/sandbox_app.py [hunk 5]
```diff
@@ -140,59 +128,64 @@ def __init__(self, args, config, gui_input):
...
+            self.habitat_env,
+            self.get_sim(),
+            lambda: self._compute_action_and_step_env(),
+            self._step_recorder,
+            lambda: self._get_recent_metrics(),
+            local_end_episode,
+            lambda: self._set_cursor_style,
+        )
+
+        if args.app_state == "fetch":
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:168]: @rpartsey @rutadesai you can create an `app_state_socialnav.py` where you implement the user busywork task we talked about (moving between objects, with help text specific to this task).

File: examples/siro_sandbox/sandbox_app.py [hunk 6]
```diff
@@ -140,59 +128,64 @@ def __init__(self, args, config, gui_input):
...
+            self._app_state_fetch = AppStateFetch(
```
Comment (rpartsey) [inline examples/siro_sandbox/sandbox_app.py:169]: Why do you. create `self._app_state_fetch` and `self._app_state_rearrange`? Can't we initialize `self._app_state` directly?

File: examples/siro_sandbox/sandbox_app.py (hunk 3 above)
Comment (rpartsey) [reply examples/siro_sandbox/sandbox_app.py:152]: My understanding is that SandboxService is the interface that should be used by (passed to) app state objects only. 

We shouldn't keep it as an instance attribute and if we need to access attributes / call methods from within  SandboxDriver we should call them directly.

File: examples/siro_sandbox/sandbox_app.py [hunk 7]
```diff
@@ -1022,95 +357,34 @@ def _record_metrics(self, metrics):
...
-            self._update_grasping_and_set_act_hints()
-
         # Navmesh visualization only works in the debug third-person view
         # (--debug-third-person-width), not the main sandbox viewport. Navmesh
         # visualization is only implemented for simulator-rendering, not replay-
         # rendering.
-        if self.gui_input.get_key_down(GuiInput.KeyNS.N):
-            self.habitat_env._sim.navmesh_visualization = (  # type: ignore
-                not self.habitat_env._sim.navmesh_visualization  # type: ignore
+        if self._sandbox_service.gui_input.get_key_down(GuiInput.KeyNS.N):
```
Comment (rpartsey) [inline examples/siro_sandbox/sandbox_app.py:376]: How about directly accessing (defining attribute in __init__ if needed) `gui_input` and `sim` from within `SandboxDriver`?

**Review by rpartsey — APPROVED** [review]
LGTM

File: examples/siro_sandbox/app_states/app_state_fetch.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+
+class AppStateFetch(AppState):
+    def __init__(
+        self,
+        sandbox_service,
+        gui_agent_ctrl,
+    ):
+        self._sandbox_service = sandbox_service
+        self.gui_agent_ctrl = gui_agent_ctrl
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_fetch.py:26]: self._gui_agent_ctrl?

File: examples/siro_sandbox/app_states/app_state_rearrange.py [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+
+
+class AppStateRearrange(AppState):
+    def __init__(
+        self,
+        sandbox_service,
+        gui_agent_ctrl,
+    ):
+        self._sandbox_service = sandbox_service
+        self.gui_agent_ctrl = gui_agent_ctrl
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_rearrange.py:26]: self._gui_agent_ctrl?

File: examples/siro_sandbox/app_states/app_state_fetch.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+        self._can_grasp_place_threshold = (
+            self._sandbox_service.args.can_grasp_place_threshold
+        )
+
+        self.cam_transform = None
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_fetch.py:32]: self._cam_transform?

File: examples/siro_sandbox/app_states/app_state_rearrange.py [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+        config = self._sandbox_service.config
+        self._end_on_success = config.habitat.task.end_on_success
+        self._obj_succ_thresh = config.habitat.task.obj_succ_thresh
+        self._success_measure_name = config.habitat.task.success_measure
+
+        self._can_grasp_place_threshold = (
+            self._sandbox_service.args.can_grasp_place_threshold
+        )
+
+        self.cam_transform = None
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_rearrange.py:38]: self._cam_transform?

File: examples/siro_sandbox/app_states/app_state_fetch.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+        self._held_target_obj_idx = None
+
+        # will be set in on_environment_reset
+        self._target_obj_ids = None
+
+        self._camera_helper = CameraHelper(
+            self._sandbox_service.args, self._sandbox_service.gui_input
+        )
+        self._first_person_mode = self._sandbox_service.args.first_person_mode
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_fetch.py:42]: We may potentially use self._camera_helper._first_person_mode (may consider making this attribute public).

File: examples/siro_sandbox/app_states/app_state_fetch.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+                walk_dir = candidate_walk_dir
+
+        self.gui_agent_ctrl.set_act_hints(
+            walk_dir,
+            grasp_object_id,
+            drop_pos,
+            self._camera_helper.lookat_offset_yaw,
+        )
+
+        return drop_pos
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_fetch.py:125]: We don't need to return `drop_pos`, the object is dropped at `drop_pos` in `self.gui_agent_ctrl.set_act_hints`.

File: examples/siro_sandbox/app_states/app_state_rearrange.py [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+                walk_dir = candidate_walk_dir
+
+        self.gui_agent_ctrl.set_act_hints(
+            walk_dir,
+            grasp_object_id,
+            drop_pos,
+            self._camera_helper.lookat_offset_yaw,
+        )
+
+        return drop_pos
```
Comment (rpartsey) [inline examples/siro_sandbox/app_states/app_state_rearrange.py:171]: We don't need to return `drop_pos`, the object is dropped at `drop_pos` in `self.gui_agent_ctrl.set_act_hints`.

File: examples/siro_sandbox/sandbox_app.py (hunk 6 above)
Comment (eundersander) [reply examples/siro_sandbox/sandbox_app.py:169]: Yes, but note that in the future, SandboxDriver will need to create multiple AppStates, e.g. AppStateTutorial and then transition to AppStateRearrange.

File: examples/siro_sandbox/app_states/app_state_fetch.py (hunk 12 above)
Comment (eundersander) [reply examples/siro_sandbox/app_states/app_state_fetch.py:42]: In my mind, these are somewhat separate. The AppState's first-person mode dictates user controls, which would probably be true even if we swapped out the camera in the future and even if we didn't have a camera.

**Pull request merged**